src/sema/eval.rs

```diff
@@ -196,3 +201,186 @@ pub fn eval_const_rational(
+            } else {
+                Expression::ShiftLeft(*loc, ty.clone(), Box::new(left), Box::new(right))
+            }
+        }
+
+        Expression::ShiftRight(loc, ty, left, right, _) => {
+            let left = eval_constants_in_expression(left, ns, results);
+            let right = eval_constants_in_expression(right, ns, results);
+
+            if let (Expression::NumberLiteral(_, _, left), Expression::NumberLiteral(_, _, right)) =
+                (&left, &right)
+            {
+                results.push(left.shr(right.to_u32().unwrap()));
+                Expression::NumberLiteral(*loc, ty.clone(), left.shr(right.to_u32().unwrap()))
+            } else {
+                Expression::ShiftLeft(*loc, ty.clone(), Box::new(left), Box::new(right))
+            }
+        }
+        Expression::NumberLiteral(_, _, n) => {
+            results.push(n.clone());
+            expr.clone()
+        }
+
+        Expression::Builtin(.., args) => {
+            for args_iter in args {
+                verify_result(args_iter, ns, &args_iter.loc());
+            }
+
+            expr.clone()
+        }
+
+        Expression::InternalFunctionCall { args, .. }
+        | Expression::ExternalFunctionCall { args, .. } => {
+            for args_iter in args.clone() {
+                verify_result(&args_iter, ns, &args_iter.loc());
+            }
+
+            expr.clone()
+        }
+        _ => expr.clone(),
+    }
+}
+
+fn overflow_check(ns: &mut Namespace, result: BigInt, ty: Type, loc: Loc) {
+    if let Type::Uint(bits) = ty {
+        // If the result sign is minus, throw an error.
+        if let Sign::Minus = result.sign() {
+            ns.diagnostics.push(Diagnostic::error(
+                loc,
+            format!( "negative value {} does not fit into type {}. Cannot implicitly convert signed literal to unsigned type.",result,ty.to_string(ns)),
+            ));
+        }
+
+        // If bits of the result is more than bits of the type, throw and error.
+        if result.bits() > bits as u64 {
+            ns.diagnostics.push(Diagnostic::error(
+                loc,
+                format!(
+                    "value {} does not fit into type {}.",
+                    result,
+                    ty.to_string(ns)
+                ),
+            ));
+        }
+    }
+
+    if let Type::Int(bits) = ty {
+        // If number of bits is more than what the type can hold. BigInt.bits() is not used here since it disregards the sign.
+        if result.to_signed_bytes_be().len() * 8 > (bits as usize) {
+            ns.diagnostics.push(Diagnostic::error(
+                loc,
+                format!(
+                    "value {} does not fit into type {}.",
+                    result,
+                    ty.to_string(ns)
+                ),
+            ));
+        }
+    }
+}
+
+pub fn verify_result(expr: &Expression, ns: &mut Namespace, loc: &Loc) {
```

Comment:
Doc comment missing.
The function should not be public. When writing code, we should limit the visibility as much as possible. Please, check if you can make this function `pub(super)` instead.